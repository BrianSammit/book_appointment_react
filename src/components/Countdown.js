import React, { Component } from 'react';
import moment, { duration } from 'moment';
import PropTypes from 'prop-types';

export default class Countdown extends Component {
  constructor(props) {
    super(props);

    this.state = {
      days: 0,
      hours: 0,
      min: 0,
      sec: 0,
    };
  }

  componentDidMount() {
    this.setCountdown();
    this.interval = setInterval(() => {
      this.setCountdown();
    }, 1000);
  }

  componentWillUnmount() {
    clearInterval(this.interval);
  }

  setCountdown() {
    const futureDate = moment(this.props.futureDate); /* eslint-disable-line */

    const today = moment();

    const clockDuration = duration(futureDate.diff(today));

    const days = Math.floor(clockDuration.asDays());
    const hours = clockDuration.hours();
    const min = clockDuration.minutes();
    const sec = clockDuration.seconds();

    this.setState({
      days,
      hours,
      min,
      sec,
    });
  }

  addZeros(value) {
    this.value = String(value);
    while (value.length < 2) {
      value = `0${value}`; /* eslint-disable-line */
    }
    return value;
  }

  render() {
    return (
      <div className="countdown">
        {Object.keys(this.state).map((key) => (
          <div className="countdown-segment" key={key.toString()}>
            {/* eslint-disable-next-line */}
            <span className="countdown-segment-number">{this.addZeros(this.state[key])}</span>
            <span className="countdown-segment-caption">{key.toUpperCase()}</span>
          </div>
        ))}
      </div>
    );
  }
}

Countdown.propTypes = {
  futureDate: PropTypes.number.isRequired,
};
