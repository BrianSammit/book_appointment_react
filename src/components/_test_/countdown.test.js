/**
 * @jest-environment jsdom
 */

import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import Countdown from '../Countdown';

describe('Creates a component', () => {
  it('Renders a component', () => {
    render(
      <Router>
        <Countdown />
      </Router>,
    );
    const element = screen.getByText('HOURS');
    expect(element).toBeInTheDocument();
  });
});
