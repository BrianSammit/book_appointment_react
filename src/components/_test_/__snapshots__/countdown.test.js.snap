// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Creates a component renders correctly 1`] = `
<div
  className="countdown"
>
  <div
    className="countdown-segment"
  >
    <span
      className="countdown-segment-number"
    >
      0
    </span>
    <span
      className="countdown-segment-caption"
    >
      DAYS
    </span>
  </div>
  <div
    className="countdown-segment"
  >
    <span
      className="countdown-segment-number"
    >
      0
    </span>
    <span
      className="countdown-segment-caption"
    >
      HOURS
    </span>
  </div>
  <div
    className="countdown-segment"
  >
    <span
      className="countdown-segment-number"
    >
      0
    </span>
    <span
      className="countdown-segment-caption"
    >
      MIN
    </span>
  </div>
  <div
    className="countdown-segment"
  >
    <span
      className="countdown-segment-number"
    >
      0
    </span>
    <span
      className="countdown-segment-caption"
    >
      SEC
    </span>
  </div>
</div>
`;
